roc/sender-receiver_test.go

```diff
@@ -0,0 +1,127 @@
+package roc
+
+import (
+	"testing"
+)
+
+func makeReceiverConfig() *ReceiverConfig {
+	return &ReceiverConfig{
+		FrameSampleRate:  44100,
+		FrameChannels:    ChannelSetStereo,
+		FrameEncoding:    FrameEncodingPcmFloat,
+		AutomaticTiming:  1,
+		ResamplerProfile: ResamplerDisable,
+	}
+}
+
+func makeSenderConfig() *SenderConfig {
+	return &SenderConfig{
+		FrameSampleRate:  44100,
+		FrameChannels:    ChannelSetStereo,
+		FrameEncoding:    FrameEncodingPcmFloat,
+		AutomaticTiming:  1,
+		ResamplerProfile: ResamplerDisable,
+		FecCode:          FecRs8m,
+	}
+}
+
+func prepare(t *testing.T) (*Sender, *Receiver, *Context) {
+	// create context
+	ctx, err := OpenContext(&ContextConfig{})
+	if ctx == nil || err != nil {
+		t.Errorf("Cannot create context, contex: %v, error: %v", ctx, err)
+	}
+
+	// create receiver
+	receiver, err := OpenReceiver(ctx, makeReceiverConfig())
+	if receiver == nil || err != nil {
+		t.Errorf("Cannot create receiver, receiver: %v, error: %v", receiver, err)
+	}
+
+	// bind receiver to two ports
+	// 1)
+	sourceAddr, err := NewAddress(AfAuto, "127.0.0.1", 0)
+	if sourceAddr == nil || err != nil {
+		t.Errorf("Cannot create source address object, address object: %v, error: %v", sourceAddr, err)
+	}
+	err = receiver.Bind(PortAudioRepair, ProtoRs8mRepair, sourceAddr)
+	if err != nil {
+		t.Errorf("Cannot bind receiver: %v", err)
+	}
+
+	// 2)
+	repairAddr, err := NewAddress(AfAuto, "127.0.0.1", 0)
+	if repairAddr == nil || err != nil {
+		t.Errorf("Cannot create repair address object, address object: %v, error: %v", repairAddr, err)
+	}
+	err = receiver.Bind(PortAudioRepair, ProtoRs8mRepair, repairAddr)
+	if err != nil {
+		t.Errorf("Cannot bind receiver: %v", err)
+	}
+
+	// create sender
+	sender, err := OpenSender(ctx, makeSenderConfig())
+	if sender == nil || err != nil {
+		t.Errorf("Cannot create sender, sender: %v, error: %v", sender, err)
+	}
+
+	// bind sender to a port
+	senderAddr, err := NewAddress(AfAuto, "127.0.0.1", 0)
+	err = sender.Bind(senderAddr)
+	if err != nil {
+		t.Errorf("Cannot bind sender: %v", err)
+	}
+
+	// connect sender to receiver ports
+	// 1)
+	err = sender.Connect(PortAudioSource, ProtoRtpRs8mSource, sourceAddr)
+	if err != nil {
+		t.Errorf("Cannot connect sender to receiver: %v", err)
+	}
+	// 2)
+	err = sender.Connect(PortAudioSource, ProtoRtpRs8mSource, repairAddr)
+	if err != nil {
+		t.Errorf("Cannot connect sender to receiver: %v", err)
+	}
+	return sender, receiver, ctx
+}
+
+func Test_roc_sender_write_receiver_read(t *testing.T) {
+	sender, receiver, ctx := prepare(t)
+
+	bytesCnt := 128
+	b := make([]byte, bytesCnt)
+	for i := 0; i < bytesCnt; i++ {
+		b[i] = byte(i)
+	}
+
+	err := sender.WriteBytes(b, 1)
+	if err != nil {
+		t.Fail()
+	}
+
+	recBytes := make([]byte, bytesCnt)
+	err = receiver.ReadAudioBytes(recBytes, 1)
```

Comment:
This piece of code has several problems:

1. Writing raw bytes is dangerous. In makeSenderConfig we set FrameEncoding to FrameEncodingPcmFloat (the only supported encoding currently), which mean that we should write floats, not bytes. We also set FrameChannels to ChannelSetStereo (the only supported channel set currently), which means that we should write samples in interleaved format, i.e. pairs of L and R samples. So the minimum unit we can pass to roc_sender_write is two floats. Floats are 32 bit, so the minimum unit is 8 bytes. And the frame size should be multiple of 8 bytes too. So our write and read calls are incorrect. I suggest to get rid of byte functions at all and use WriteFloats/ReadFloats here.

2. The stream read from receiver is not guaranteed to be identical to the stream written to sender. First, received stream will have some leading zeros, not present in sent stream, because of the jitter buffer latency. Second, received stream may contain zeros instead of some samples if corresponding packet was lost and receiver wasn't able to recover it (or FEC is disabled). Third, received stream may contain shifts, if the session was restarted in the middle of a stream, which happens e.g. on high loss rate. This is not a frequent event in real use cases, however it happens from time to time, including travis workers which are sometimes very loaded and thus slow and thus packet loss increases. Forth, if resampling is enabled, sample rate conversion may change actual sample values a bit. The difference is not audible, but the samples are not equal numbers any more even if there is no packet loss.

3. Sender and receiver streams are both real-time and they don't block like in TCP. Receiver read does not wait until something is received. If some samples were not received in time, they're just replaced with zeros. If no samples were received, or there are no senders at all, receiver just produces all zeros. So if you call receiver read in a loop and write a single non-zero sample to sender (say, from another thread), receiver will produce a stream of zeros with a single non-zero sample *somewhere in the stream*. It's not guaranteed to be the first sample you'll get from receiver. And it's not guaranteed to be present at all because we use UDP, which is unreliable. See also my comment here.

Reply:
The idea of libroc is to provide a simple TCP-like API and hide the complexity of real-time streaming behind it.

The user can just write samples, read samples, and libroc will try to fulfill two contradicting requirements: 1) maintain good service quality 2) maintain fixed latency. In other words, it will try to "sound good". But that's all. We don't (and can't) provide stronger guarantees like in TCP. This is the cost we pay for real-time.

So the API is TCP-like but the behavior is not. You can expect that it will sound good, but you can't make much assumptions of the exact content on the received stream.

BTW, in future we likely will add more transports, including reliable ones like HTTP. But so far this is not our major use case.

Reply:
To make this test reliable, we need to send samples in a loop and read samples in a loop, and wait until we receive at least N non-zero samples. The same technique was implemented in https://github.com/roc-project/roc/pull/283. Please take a look at that issue and discussions in the PR, in particular this one: https://github.com/roc-project/roc/pull/283#discussion_r341954058

Reply:
Yes, and regarding the test here: https://github.com/roc-project/roc-tests/blob/master/tests/test_service_quality.cpp

roc-tests repo is intended to collect *real-time* tests. They wont be 100% reliable. Their result wont be just boolean like "passed" or "failed", but instead something like "the loss ration was below 5% this time". And they will be run on real hardware on unloaded system.

We can't use such approach for our regular unit and integration tests because they are run on random machines including slow travis worker, under valgrind on with sanitizers that slow down execution, etc. These tests should be 100% reliable and thus their requirements should be less strict.

Reply:
> Please take a look at the related issue and discussions in the PR, in particular this one: roc-project/roc#283 (comment)

alright, i will take a look at more relaxed tests, thanks for the elaborated hints!

Reply:
I think something went wrong with github pages after rename, at list I'm not able to go to this link https://roc-streaming.org/roc/docs/internals/data_flow.html

Reply:
The new URL is https://roc-streaming.org/toolkit/docs/internals/data_flow.html

(/roc/ was changed to /toolkit/)

Current website still have some broken URLs. I'll fix them during next few days. If you'll find broken URLs later than in a few day, please report them to me :)